src/System.Memory/src/System/Buffers/ArrayBufferWriter.cs

```diff
@@ -0,0 +1,223 @@
+            _totalWritten += _written;
+
+            ResetHelper();
+        }
+
+        public void CopyToAndReset(Stream stream)
+        {
+            CheckIfDisposed();
+
+            if (stream == null)
+                throw new ArgumentNullException(nameof(stream));
+
+            stream.Write(_rentedBuffer, 0, _written);
+            _totalWritten += _written;
+
+            ResetHelper();
+        }
+
+        // Returns the rented buffer back to the pool
+        public void Dispose()
+        {
+            if (_rentedBuffer == null)
+            {
+                return;
+            }
+
+            ArrayPool<byte>.Shared.Return(_rentedBuffer, clearArray: true);
+            _rentedBuffer = null;
+            _written = 0;
+            _totalWritten = 0;
+        }
+
+        private void CheckIfDisposed()
+        {
+            if (_rentedBuffer == null)
+                throw new ObjectDisposedException(nameof(ArrayBufferWriter));
+        }
+
+        void IBufferWriter<byte>.Advance(int count)
+        {
+            CheckIfDisposed();
+
+            if (count < 0)
+                throw new ArgumentException(nameof(count));
+
+            if (_written > _rentedBuffer.Length - count)
+                ThrowHelper.ThrowInvalidOperationException(_rentedBuffer.Length);
+
+            _written += count;
+        }
+
+        Memory<byte> IBufferWriter<byte>.GetMemory(int sizeHint)
+        {
+            CheckIfDisposed();
+
+            if (sizeHint < 0)
+                throw new ArgumentException(nameof(sizeHint));
+
+            CheckAndResizeBuffer(sizeHint);
+            return _rentedBuffer.AsMemory(_written);
+        }
+
+        Span<byte> IBufferWriter<byte>.GetSpan(int sizeHint)
+        {
+            CheckIfDisposed();
+
+            if (sizeHint < 0)
+                throw new ArgumentException(nameof(sizeHint));
+
+            CheckAndResizeBuffer(sizeHint);
+            return _rentedBuffer.AsSpan(_written);
+        }
+
+        private void CheckAndResizeBuffer(int sizeHint)
+        {
+            Debug.Assert(sizeHint >= 0);
+
+            if (sizeHint == 0)
+            {
+                sizeHint = MinimumBufferSize;
+            }
+
+            int availableSpace = _rentedBuffer.Length - _written;
+
+            if (sizeHint > availableSpace)
+            {
+                int growBy = sizeHint > _rentedBuffer.Length ? sizeHint : _rentedBuffer.Length;
+
+                int newSize = checked(_rentedBuffer.Length + growBy);
+
+                byte[] oldBuffer = _rentedBuffer;
+
+                _rentedBuffer = ArrayPool<byte>.Shared.Rent(newSize);
+
+                Debug.Assert(oldBuffer.Length >= _written);
+                Debug.Assert(_rentedBuffer.Length >= _written);
+
+                oldBuffer.AsSpan(0, _written).CopyTo(_rentedBuffer);
+                ArrayPool<byte>.Shared.Return(oldBuffer, clearArray: true);
```

Comment:
We may not have written the full buffer. How about clearing just what we wrote?

Reply:
Should we do the same thing on Dispose() or is it fine to clear the whole array (given we only do it once)?